Graphics/X11/Xft.hsc

```diff
@@ -219,6 +249,58 @@ xftTextExtents d f string =
       cXftTextExtentsUtf8 d f str_ptr (fi len) cglyph
       peek cglyph
 
+-- Support for multiple fonts --
+
+-- | Like 'xftDrawString', but fall back to another font in the given
+-- list if necessary (i.e., should a character not be drawable with the
+-- currently selected font).
+xftDrawStringFallback :: XftDraw -> XftColor -> [XftFont] -> Int -> Int -> String -> IO ()
+xftDrawStringFallback d c fs x y string = do
+    display <- xftDrawDisplay d
+    chunks <- getChunks display fs string
+    mapM_ (\(f, s, _, xo, yo) -> xftDrawString d c f (x + xo) (y + yo) s) chunks
+
+-- | Like 'xftTextExtents' but for multiple fonts.  Return the extents
+-- for all fonts that are needed to draw the given string.
+xftTextExtentsChunks :: Display -> [XftFont] -> String -> IO [XGlyphInfo]
+xftTextExtentsChunks d fs string = map (\(_, _, gi, _, _) -> gi) <$> getChunks d fs string
```

Comment:
I think this may be a bit confusing since we could have the same glyphinfo in multiple places in that list (think "string *unicode_symbol* string") but no indication of which parts of the text it belongs to.  I suppose we could either return something like `[(String, XGlyphInfo)]` or we could nub the list (probably easiest by deriving `Eq` on `XftFont` and then calling `nubBy` before `map`).

Reply:
I just mention that `getOffsets` in `getChunks` actually computes wrong `XGlyphInfo` for each independent substring. It accumulates the width and height for whatever reason. Or maybe I misunderstood the semantic of `XGlyphInfo` fields. So I think I could work on that, and then just return `[(String, XGlyphInfo)]` in `xftTextExtentsChunks`

Reply:
Well, font fallback for xmobar certainly works, which makes this very weird indeed.  I'm not so sure about what `XGlyphInfo` is really supposed to represent either—documentation seems rather sparse.

@jaor since this code originally came from xmobar (and according to the module docs from you!) can you shed any light on this?  I hope it's not too indecent to ask about code from like 7 years ago :)

Reply:
In xmbar it makes sense, while `xftTextEtentsChunks` returns the last, which contains info about all chunks in one such struct

Reply:
Ah indeed, I forgot about `xftDrawStringFallback` not using this information at all.  I guess what we should really be doing then is to return something like `IO (XGlyphInfo, [(XftFont, String, Int, Int)])`, right?

Reply:
Actually, we can make two separate ones, `xftTextExtentsChunks` to return `[(String, XftFont, XGlyphInfo)]` (we do not xo, yo in this tuple because we'll store it in `XGlyphInfo`) and something like `xftTextAccumulativeExtents` to return just `XGlyphInfo` like xmobar does for the main use case

Reply:
At this point we may well just drop the idea of `xftTextExtentsChunks` and just expose `xftTextAccumulativeExtents`.  I'm not sure the fomer has any uses, I just immediately thought of it when I saw that call to `last`

Reply:
@iliayar LGTM, feel free to squash this down and then we can merge